---
title: "Post_taxonomy_Pipeline_ITS (2/2)"
author: "Chung-Wing Ko"
date: "1 Feb 2024"
output: 
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    code_folding: show
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

# Set up

This file is for post-asv file generation. Next steps include including functional trait database, creating a phyloseq object, filtering out moulds, and rarefaction.

```{r}
#install packages
#BiocManager::install("microbiome")
#devtools::install_github("KarstensLab/microshades")
#BiocManager::install("miaViz")
#install.packages("remotes")
#remotes::install_github("david-barnett/microViz")
#devtools::install_github("vmikk/metagMisc")

#load libraries
library(tidyr)
library(dplyr)
library(readr)
library(tidyverse)
library(phyloseq)
library(vegan)
library(ggplot2)
library(gam)
library(BiodiversityR)
library(microbiome)
library(microshades)
library(miaViz)
library(microViz)
library(phyloseq); packageVersion("phyloseq")
library(metagMisc); packageVersion("metagMisc")

#set working directory
setwd('/Users/chungwingko/Desktop/SCELSCE_Sequencing/ITS_Trial/Post_phyloseq')
path <- setwd('/Users/chungwingko/Desktop/SCELSCE_Sequencing/ITS_Trial/Post_phyloseq')
list.files(path)
```

# Read in data

These files are ASV files post-taxonomy processing. Three main files to focus on: taxonomy (can filter out all non-fungal species) which will be combined with functional data, count data (abundances), and environmental metadata (your own dataset). 

```{r}
#this is looking at all eukaryotes
#asv_tax_counts_all <- read_tsv("/Users/chungwingko/Desktop/SCELSCE_Sequencing/ITS_Trial/Post_phyloseq/ASVs_tax_counts.tsv")
#compared to just fungi -- seems like 1/3 of results are not fungi??
asv_tax_counts <- read_tsv("/Users/chungwingko/Desktop/SCELSCE_Sequencing/ITS_Trial/Post_phyloseq/ASVs_tax_counts_only_fungi.tsv")
asv_tax_counts <- asv_tax_counts[,-1]

#taxonomy
asv_tax <- asv_tax_counts %>% select(c(1:8))
asv_tax <- asv_tax %>% rename_at(1,~"asv")
asv_tax <- asv_tax %>% remove_rownames %>% column_to_rownames(var = "asv")
Genus<-gsub("g__","",asv_tax$Genus)
write.csv(asv_tax, "ASV_Tax.csv")

#counts
asv_tab <- asv_tax_counts %>% select(c(1, 9:24))
asv_tab <- asv_tab %>% rename_at(1,~"asv")
asv_tab <- asv_tab %>% remove_rownames %>% column_to_rownames(var = "asv")
write.csv(asv_tab, "ASV_Abund.csv")

#environ data
env_data <- read.csv("/Users/chungwingko/Desktop/SCELSCE_Sequencing/ITS_Trial/Post_phyloseq/Soil_data_trial_FT.csv")
env_data <- env_data %>% remove_rownames %>% column_to_rownames(var = "Sample")
env_data$Forest <- as.factor(env_data$Forest)
write.csv(env_data, "Environ_Data.csv")
```

# Functional Database

Remember there are limitations in assumptions cause we're just noting organisms rather than activity, and all of this is entirely based on what's already known. Different tools are also based on diff levels of taxonomical classification, so important to read documentation. 

There are two main tools: 

## 1. FunGuild

Link: *https://github.com/UMNFuN/FUNGuild*
Link: *https://github.com/brendanf/FUNGuildR*
These are R packages but can work outside R also. 

Essentially, we're combining all fungal traits with our taxonomy and abundance (matching tables based only on taxonomical names). 

```{r, eval=FALSE}
#set up
library(tidyverse); packageVersion('tidyverse')
#devtools::install_github("brendanf/FUNGuildR")
library(FUNGuildR)
```

#### Organising

```{r}
# removing all the prefixes built into our table so that the taxonomic names can match
asv_tax_counts2 <- data.frame(lapply(asv_tax_counts, function(x) {gsub('k__', '', x)}))
asv_tax_counts2 <- data.frame(lapply(asv_tax_counts2, function(x) {gsub('p__', '', x)}))
asv_tax_counts2 <- data.frame(lapply(asv_tax_counts2, function(x) {gsub('c__', '', x)}))
asv_tax_counts2 <- data.frame(lapply(asv_tax_counts2, function(x) {gsub('o__', '', x)}))
asv_tax_counts2 <- data.frame(lapply(asv_tax_counts2, function(x) {gsub('f__', '', x)}))
asv_tax_counts2 <- data.frame(lapply(asv_tax_counts2, function(x) {gsub('g__', '', x)}))
asv_tax_counts2 <- data.frame(lapply(asv_tax_counts2, function(x) {gsub('s__', '', x)}))

#species name has two names, species and epithet -- need to combine genus and species to recognise in database
#Species_full is a new column containing full name of genus species
asv_tax_counts2.tax <- unite(asv_tax_counts2, Species_full, Genus, Species, sep = ' ', na.rm = TRUE, remove = FALSE)

#create another column combining all the taxonomy -- requires one column with all taxonomy
asv_tax_counts2.tax <- unite(asv_tax_counts2.tax, Taxonomy, Kingdom, Phylum, Class, Order, Family, Genus, Species_full, sep = ';', na.rm = TRUE, remove = TRUE)

#then remove species -- finalised table is ready to go!
asv_tax_counts2.tax <- asv_tax_counts2.tax[,-3]

#save file
saveRDS(asv_tax_counts2.tax, "asv_tax_counts_clean.Rds")
```

#### FUNGuild

```{r, eval = FALSE}
#store the FUNGuild database
fun_db <- get_funguild_db(db = "http://www.stbates.org/funguild_db_2.php")
#this takes a little while

#run FUNGuild
its_fun <- funguild_assign(asv_tax_counts_rarefied2_tax, db = fun_db, tax_col = 'Taxonomy') #specify which column the taxonomy is in

its_fun
write.table(its_fun, 'ITS_FUNGuild.tsv', sep = '\t', quote = F, row.names = F)

its_fun_filtered <- drop_na(its_fun, trophicMode)
write.table(its_fun_filtered, 'ITS_FUNGuild_filtered.tsv', sep = '\t', quote = F, row.names = F)
```


Looking at table: trophic mode, guild, probability (this is qualitative, not CIs). They're having an update later this year (2024), check this closely for timeline. 

Remember that not all matches will have full information provided. Keep track of every step in your own notes, so that when (if) you write a paper, you can take note (like we are doing taxonomic analysis on XX amount of samples, out of this we have functional data for XX% of these samples, so here is the functional guild breakdown for the ones we know). 


## 2. FungalTraits

This is a second, different database. This is generally considered more complete, and Kabir uses this, but FunGuild is scheduled to have a major update, so is worth keeping an eye on. Different databases may be better for different groups (differences in how they name groups and functions, stronger foundations in different sets of literature, and used for different end goals), so you should look at literature to see which is more commonly used for your purpose/more updated (so you should be able to justify why you chose the one you chose).

FungalTraits is very well published and used, but has less of a github presence (more of like an excel file) so a bit harder to keep track of versions, and less documented updates. 

```{r}
#import database as a csv file
fungal.traits.database <- read.csv('/Users/chungwingko/Desktop/SCELSCE_Sequencing/ITS_Trial/Post_phyloseq/FungalTraits 1.2_ver_16Dec_2020.csv')
fungal.traits.database <- fungal.traits.database %>% dplyr::rename(Genus = GENUS)

#then matching genera to database
#funguild has their built in functions, this database works manually
traits_table <- fungal.traits.database[match(Genus,fungal.traits.database$Genus),]
tax.trait.table <- cbind(asv_tax_counts2.tax, traits_table)
tax.trait.table <- tax.trait.table %>% rename_at(1,~"asv")
tax.trait.table <- tax.trait.table %>% remove_rownames %>% column_to_rownames(var = "asv")

#save
saveRDS(tax.trait.table, "ITS_Post_FungalTraits")
```

## Phyloseq object

```{r}
#read in table
tax.trait.table <- readRDS("/Users/chungwingko/Desktop/SCELSCE_Sequencing/ITS_Trial/Post_phyloseq/ITS_Post_FungalTraits")

#create phyloseq object using ASV files
ITS_phyloseq <- phyloseq(otu_table(asv_tab, taxa_are_rows=TRUE), sample_data(env_data), tax_table(as.matrix(tax.trait.table))) # taxa in rows
ITS_phyloseq
saveRDS(ITS_phyloseq, "ITS_Phyloseq.Rds")
```

## Filter out moulds

```{r}
# new phyloseq w filtered (no aspergillaceae)
ITSphyloseq.filtered <- subset_taxa(ITS_phyloseq, Family!="f__Aspergillaceae")

saveRDS(ITSphyloseq.filtered, "ITS_phyloseq_filtered.Rds")
```

# Rarefaction

To decrease bias, we rarefy data to have the same number of reads across all samples. This step takes the sample with the least number of reads and randomly draws reads from all other samples up until that threshold, then caps that number. 

Why do we rarefy? 
Based on the quality filtering, there will be different amounts of reads for each resulting sample. This creates unequal sampling, and can show false species richness and diversity across samples which would be a result of uneven sample numbers, rather than true ecological differences. 

What potential issues are there with rarefaction: 
1. We are losing data that we paid for, so throwing away a lot of money.
2. Bias against rarer species and losing diversity. If using rare data, don't want to rarify, use relative abundance instead. *Note: Working with proportions rather than absolute numbers and abundance is also problematic, as most stats tests work with absolute numbers instead*
3. If some samples are really low quality, better to remove those few samples rather than sequences from the rest. Can plot a histogram of the sample sums and see if a few samples are dragging the rest down. 

## Checking range, setting minimum

```{r}
#check range for reads, take minimum
range(sample_sums(ITSphyloseq.filtered))
mean(sample_sums(ITSphyloseq.filtered)) #mean
hist(sample_sums(ITSphyloseq.filtered)) #plot distribution

#another way: mean(as.data.frame(track)$nonchim); min(as.data.frame(ITSphyloseq.filtered)$nonchim)
```

## Rarefy

```{r}
#set sample size as minimum amt, random process, normally want at least 5-10k 
physeq_rarefied <- rarefy_even_depth(ITSphyloseq.filtered, rngseed = TRUE, replace = FALSE, sample.size = 26634)
physeq_rarefied
#checking that the sequence #s are now the same
sample_sums(physeq_rarefied)

#save
saveRDS(physeq_rarefied,"physeq_cleaned_ITS")

#making the rarefied data in a table

##abundance data only
asv_tab_rarefied <- as(otu_table(physeq_rarefied), "matrix")
apply(asv_tab_rarefied, 2, sum)
write.table(asv_tab_rarefied, "ASVs_counts_rarefied.tsv", sep="\t", quote=F, col.names=NA)

##taxonomy and abundance
asv_tax_counts_rarefied <- phyloseq_to_df(physeq_rarefied, addtax = T, addtot = F, sorting = "NULL")

#to double check the rarefaction worked
apply(asv_tax_counts_rarefied[,9:10], 2, sum)
write.table(asv_tax_counts_rarefied, "ASVs_tax_counts_rarefied.tsv", sep="\t", quote=F, col.names=NA)

#save files (never overwrite files!!)
saveRDS(asv_tax_counts_rarefied, "asv_tax_counts.Rds")
#remember to note all versions and every step for publishing later
```


```{r}
# kabir's code for rarefaction
# physeq.even <-rarefy_even_depth(physeq,sample.size=50000,rngseed=22)
```

